server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java

```diff
@@ -165,34 +166,61 @@ public void insert(
   {
     try {
       getConnector().retryWithHandle(
-          (HandleCallback<Void>) handle -> {
-            final String sql = StringUtils.format(
-                "INSERT INTO %s (id, created_date, datasource, payload, type, group_id, active, status_payload) "
-                + "VALUES (:id, :created_date, :datasource, :payload, :type, :group_id, :active, :status_payload)",
-                getEntryTable()
-            );
-            handle.createStatement(sql)
-                  .bind("id", id)
-                  .bind("created_date", timestamp.toString())
-                  .bind("datasource", dataSource)
-                  .bind("payload", jsonMapper.writeValueAsBytes(entry))
-                  .bind("type", type)
-                  .bind("group_id", groupId)
-                  .bind("active", active)
-                  .bind("status_payload", jsonMapper.writeValueAsBytes(status))
-                  .execute();
-            return null;
-          },
-          e -> getConnector().isTransientException(e) && !(isStatementException(e) && getEntry(id).isPresent())
+          handle -> insertEntryWithHandle(handle, id, timestamp, dataSource, entry, active, status, type, groupId),
+          this::isTransientDruidException
       );
     }
+    catch (CallbackFailedException e) {
+      propageAsRuntimeException(e.getCause());
+    }
     catch (Exception e) {
-      if (isStatementException(e) && getEntry(id).isPresent()) {
-        throw new EntryExistsException(id, e);
-      } else {
-        Throwables.propagateIfPossible(e);
-        throw new RuntimeException(e);
-      }
+      propageAsRuntimeException(e);
+    }
+  }
+
+  private void propageAsRuntimeException(Throwable t)
+  {
+    Throwables.propagateIfPossible(t);
+    throw new RuntimeException(t);
+  }
+
+  /**
+   * Inserts the given entry into the metadata store. Any exception thrown by a
+   * HandleCallback is wrapped in a {@link DruidException} wrapped in a
+   * {@link CallbackFailedException}.
+   */
+  private int insertEntryWithHandle(
```

Comment:
Nit: It doesn't seem like the return value is being used. Can this be Void instead?